GWT cannot optimize switched expressions

**GWT version:** HEAD-SNAPSHOT
**Java Version:** 17

---

##### Description

For some reason GWT cannot compile the following code

##### Steps to reproduce

```java
public class Basic implements EntryPoint {
  @Override
  public void onModuleLoad() {
    getKind(Kind.A);
  }

  private static String getKind(Kind kind) {
    return switch (kind) {
      case A -> "1";
      case B -> "2";
      case C -> "3";
      case D -> "4";
      case E -> "5";
      case F -> "6";
      case G -> "7";
      case H -> "8";
      case I -> "9";
      case J -> "10";
      case K -> "11";
      case L -> "12";
      case M -> "13";
      case N -> "14";
      case O -> "15";
    };
  }
}
```
##### Output

```
Compiling module it.pkg.Basic
   Compiling 2 permutations
      Compiling permutation 0...
      [ERROR] An internal compiler exception occurred
com.google.gwt.dev.jjs.InternalCompilerException: Unexpected error during visit.
	at com.google.gwt.dev.jjs.ast.JVisitor.translateException(JVisitor.java:111)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:130)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:122)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:118)
	at com.google.gwt.dev.jjs.impl.CloneExpressionVisitor.cloneExpression(CloneExpressionVisitor.java:78)
	at com.google.gwt.dev.jjs.impl.MethodInliner$InliningVisitor.extractExpressionsFromBody(MethodInliner.java:265)
	at com.google.gwt.dev.jjs.impl.MethodInliner$InliningVisitor.tryInlineMethodCall(MethodInliner.java:147)
	at com.google.gwt.dev.jjs.impl.MethodInliner$InliningVisitor.endVisit(MethodInliner.java:112)
	at com.google.gwt.dev.jjs.ast.JMethodCall.traverse(JMethodCall.java:268)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:118)
	at com.google.gwt.dev.jjs.ast.JExpressionStatement.traverse(JExpressionStatement.java:42)
	at com.google.gwt.dev.jjs.ast.JModVisitor$ListContext.traverse(JModVisitor.java:88)
	at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemove(JModVisitor.java:331)
	at com.google.gwt.dev.jjs.ast.JBlock.traverse(JBlock.java:94)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:139)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:135)
	at com.google.gwt.dev.jjs.ast.JMethodBody.traverse(JMethodBody.java:83)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
	at com.google.gwt.dev.jjs.ast.JMethod.visitChildren(JMethod.java:786)
	at com.google.gwt.dev.jjs.ast.JMethod.traverse(JMethod.java:778)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
	at com.google.gwt.dev.jjs.impl.FullOptimizerContext.traverse(FullOptimizerContext.java:224)
	at com.google.gwt.dev.jjs.impl.MethodInliner.execImpl(MethodInliner.java:611)
	at com.google.gwt.dev.jjs.impl.MethodInliner.exec(MethodInliner.java:588)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.optimizeJavaOneTime(JavaToJavaScriptCompiler.java:1514)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.optimizeJavaToFixedPoint(JavaToJavaScriptCompiler.java:1443)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.optimizeJava(JavaToJavaScriptCompiler.java:525)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:362)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:274)
	at com.google.gwt.dev.CompilePerms.compile(CompilePerms.java:198)
	at com.google.gwt.dev.ThreadedPermutationWorkerFactory$ThreadedPermutationWorker.compile(ThreadedPermutationWorkerFactory.java:50)
	at com.google.gwt.dev.PermutationWorkerFactory$Manager$WorkerThread.run(PermutationWorkerFactory.java:74)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.UnsupportedOperationException: switch expression cannot be cloned
	at com.google.gwt.dev.jjs.impl.CloneExpressionVisitor.visit(CloneExpressionVisitor.java:261)
	at com.google.gwt.dev.jjs.ast.JSwitchExpression.traverse(JSwitchExpression.java:50)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:127)
	... 40 more
         [ERROR] at Basic.java(12): switch (Enum.$ordinal(kind))  {
  case Enum.$ordinal(Kind.A):
  yield  "1";
  case Enum.$ordinal(Kind.B):
  yield  "2";
  case Enum.$ordinal(Kind.C):
  yield  "3";
  case Enum.$ordinal(Kind.D):
  yield  "4";
  case Enum.$ordinal(Kind.E):
  yield  "5";
  case Enum.$ordinal(Kind.F):
  yield  "6";
  case Enum.$ordinal(Kind.G):
  yield  "7";
  case Enum.$ordinal(Kind.H):
  yield  "8";
  case Enum.$ordinal(Kind.I):
  yield  "9";
  case Enum.$ordinal(Kind.J):
  yield  "10";
  case Enum.$ordinal(Kind.K):
  yield  "11";
  case Enum.$ordinal(Kind.L):
  yield  "12";
  case Enum.$ordinal(Kind.M):
  yield  "13";
  case Enum.$ordinal(Kind.N):
  yield  "14";
  case Enum.$ordinal(Kind.O):
  yield  "15";
}
            com.google.gwt.dev.jjs.ast.JSwitchExpression
         [ERROR] at Basic.java(8): Basic.getKind(Kind.A)
            com.google.gwt.dev.jjs.ast.JMethodCall
         [ERROR] at Basic.java(8): Basic.getKind(Kind.A)
            com.google.gwt.dev.jjs.ast.JExpressionStatement
         [ERROR] at Basic.java(7): {
  Basic.getKind(Kind.A);
}
            com.google.gwt.dev.jjs.ast.JBlock
         [ERROR] at Basic.java(7): {
  Basic.getKind(Kind.A);
}
            com.google.gwt.dev.jjs.ast.JMethodBody
         [ERROR] at Basic.java(7): it.pkg.client.Basic.onModuleLoad()V
            com.google.gwt.dev.jjs.ast.JMethod
      [ERROR] Unrecoverable exception, shutting down
com.google.gwt.core.ext.UnableToCompleteException: (see previous log entries)
	at com.google.gwt.dev.javac.CompilationProblemReporter.logAndTranslateException(CompilationProblemReporter.java:106)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:461)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:274)
	at com.google.gwt.dev.CompilePerms.compile(CompilePerms.java:198)
	at com.google.gwt.dev.ThreadedPermutationWorkerFactory$ThreadedPermutationWorker.compile(ThreadedPermutationWorkerFactory.java:50)
	at com.google.gwt.dev.PermutationWorkerFactory$Manager$WorkerThread.run(PermutationWorkerFactory.java:74)
	at java.base/java.lang.Thread.run(Thread.java:840)
      [ERROR] Not all permutation were compiled , completed (0/2)
```

##### Known workarounds

`-optimize=0`

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/ImplementRecordComponents.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/ImplementRecordComponents.java
@@ -193,13 +193,16 @@ public class ImplementRecordComponents {
                   myField,
                   otherField);
         } else {
-          // we would like to use Objects.equals here to be more consise, but we would need
+          // We would like to use Objects.equals here to be more concise, but we would need
           // to look up the right impl based on the field - just as simple to insert a null check
           // and get it a little closer to all being inlined away
+
+          // Make another field ref to call equals() on
+          JFieldRef myField2 = new JFieldRef(info, new JThisRef(info, type), field, type);
           equals = new JBinaryOperation(info, JPrimitiveType.BOOLEAN, JBinaryOperator.AND,
                   new JBinaryOperation(info, JPrimitiveType.BOOLEAN, JBinaryOperator.NEQ,
                           myField, JNullLiteral.INSTANCE),
-                  new JMethodCall(info, myField, objectEquals, otherField));
+                  new JMethodCall(info, myField2, objectEquals, otherField));
         }
         if (componentCheck != JBooleanLiteral.TRUE) {
           componentCheck = new JBinaryOperation(info, JPrimitiveType.BOOLEAN,

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/MethodInliner.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/MethodInliner.java
@@ -288,10 +288,11 @@ public class MethodInliner {
         } else if (stmt instanceof JReturnStatement) {
           JReturnStatement returnStatement = (JReturnStatement) stmt;
           JExpression expr = returnStatement.getExpr();
-          if (!CannotBeInlinedVisitor.check(expr)) {
-            return null;
-          }
+
           if (expr != null) {
+            if (!CannotBeInlinedVisitor.check(expr)) {
+              return null;
+            }
             JExpression clone = cloner.cloneExpression(expr);
             clone = maybeCast(clone, body.getMethod().getType());
             expressions.add(clone);

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/MethodInliner.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/MethodInliner.java
@@ -76,6 +76,31 @@ public class MethodInliner {
   }
 
   /**
+   * Determines if the given expression can be inlined. Any switch expression will fail this check.
+   */
+  private static class CannotBeInlinedVisitor extends JVisitor {
+    private boolean succeed = true;
+    public static boolean check(JExpression expr) {
+      CannotBeInlinedVisitor v = new CannotBeInlinedVisitor();
+      v.accept(expr);
+      return v.succeed;
+    }
+
+    @Override
+    public boolean visit(JStatement x, Context ctx) {
+      // To ensure we didn't miss an important case, throw if we see a statement, as those cannot
+      // be inlined.
+      throw new IllegalStateException("Should never visit statements");
+    }
+
+    @Override
+    public boolean visit(JSwitchExpression x, Context ctx) {
+      succeed = false;
+      return false;
+    }
+  }
+
+  /**
    * Method inlining visitor.
    */
   private class InliningVisitor extends JChangeTrackingVisitor {
@@ -148,6 +173,7 @@ public class MethodInliner {
       if (expressions == null) {
         // If it will never be possible to inline the method, add it to a
         // blacklist
+
         return InlineResult.BLACKLIST;
       }
 
@@ -242,6 +268,9 @@ public class MethodInliner {
           if (initializer == null) {
             continue;
           }
+          if (!CannotBeInlinedVisitor.check(initializer)) {
+            return null;
+          }
           JLocal local = (JLocal) declStatement.getVariableRef().getTarget();
           JExpression clone = new JBinaryOperation(stmt.getSourceInfo(), local.getType(),
               JBinaryOperator.ASG,
@@ -251,9 +280,7 @@ public class MethodInliner {
         } else if (stmt instanceof JExpressionStatement) {
           JExpressionStatement exprStmt = (JExpressionStatement) stmt;
           JExpression expr = exprStmt.getExpr();
-          if (expr instanceof JSwitchExpression) {
-            // Switch expressions can't be cloned in this way, though we wouldn't want to inline
-            // such a large block anyway.
+          if (!CannotBeInlinedVisitor.check(expr)) {
             return null;
           }
           JExpression clone = cloner.cloneExpression(expr);
@@ -261,6 +288,9 @@ public class MethodInliner {
         } else if (stmt instanceof JReturnStatement) {
           JReturnStatement returnStatement = (JReturnStatement) stmt;
           JExpression expr = returnStatement.getExpr();
+          if (!CannotBeInlinedVisitor.check(expr)) {
+            return null;
+          }
           if (expr != null) {
             JExpression clone = cloner.cloneExpression(expr);
             clone = maybeCast(clone, body.getMethod().getType());

--- a/user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java
+++ b/user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java
@@ -439,4 +439,32 @@ public class Java17Test extends GWTTestCase {
             : 4.0;
     assertTrue(notCalled);
   }
+
+  public void testSwitchExprInlining() {
+    enum HasSwitchMethod {
+      A, RED, SUNDAY, JANUARY, ZERO;
+      public static final int which(HasSwitchMethod whichSwitch) {
+        return switch(whichSwitch) {
+          case A -> 1;
+          case RED -> 2;
+          case SUNDAY -> 3;
+          case JANUARY -> 4;
+          case ZERO -> 5;
+        };
+      }
+      public static final int pick(HasSwitchMethod whichSwitch) {
+        return 2 * switch(whichSwitch) {
+          case A -> 1;
+          case RED -> 2;
+          case SUNDAY -> 3;
+          case JANUARY -> 4;
+          case ZERO -> 5;
+        };
+      }
+    }
+
+    HasSwitchMethod uninlinedValue = Math.random() > 2 ? HasSwitchMethod.A : HasSwitchMethod.RED;
+    assertEquals(2, HasSwitchMethod.which(uninlinedValue));
+    assertEquals(4, HasSwitchMethod.pick(uninlinedValue));
+  }
 }

--- a/user/test/com/google/gwt/dev/jjs/test/Java17Test.java
+++ b/user/test/com/google/gwt/dev/jjs/test/Java17Test.java
@@ -114,6 +114,10 @@ public class Java17Test extends GWTTestCase {
     assertFalse(isGwtSourceLevel17());
   }
 
+  public void testSwitchExprInlining() {
+    assertFalse(isGwtSourceLevel17());
+  }
+
   private boolean isGwtSourceLevel17() {
     return JUnitShell.getCompilerOptions().getSourceLevel().compareTo(SourceLevel.JAVA17) >= 0;
   }

